src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -173,15 +179,159 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
                         if (InitializationCacheExpires < DateTimeOffset.UtcNow)
                         {
                             InitializationCacheExpires = DateTimeOffset.UtcNow.Add(MinCacheExpirationInterval);
-                            await LoadAll(ignoreFailures: false, cancellationToken).ConfigureAwait(false);
+                            await LoadAllWithFailOverPolicyAsync(ignoreFailures: false, cancellationToken).ConfigureAwait(false);
                         }
 
                         return;
                     }
 
-                    await RefreshIndividualKeyValues(cancellationToken).ConfigureAwait(false);
-                    await RefreshKeyValueCollections(cancellationToken).ConfigureAwait(false);
-                    await RefreshKeyValueAdapters(cancellationToken).ConfigureAwait(false);
+                    //
+                    // Avoid instance state modification
+                    Dictionary<string, ConfigurationSetting> applicationSettings = null;
+                    Dictionary<KeyValueWatcher, KeyValueChange> keyValueChanges = null;
+                    List<KeyValueChange> changedKeyValuesCollection = null;
+                    bool refreshAll = false;
+
+                    await ExecuteWithFailOverPolicyAsync(async (client) =>
+                        {
+                            keyValueChanges = new Dictionary<KeyValueWatcher, KeyValueChange>();
+                            refreshAll = false;
+                            changedKeyValuesCollection = null;
+                            applicationSettings = null;
+
+                            foreach (KeyValueWatcher changeWatcher in _options.ChangeWatchers)
+                            {
+                                // Skip the refresh for this key if the cached value has not expired
+                                if (DateTimeOffset.UtcNow < changeWatcher.CacheExpires)

```

Comment:
I think we should get the list of expired watchers outside of the failover code, so 
- we don't do anything if no watcher expired. This is very hot code path.
- we get UtcNow once regardless how many failovers we have. It helps the consistency when we fail over from one client to another.

Reply:
That's a great optimization opportunity. Done!